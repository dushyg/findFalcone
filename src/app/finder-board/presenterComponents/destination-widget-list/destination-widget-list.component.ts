import {
  Component,
  OnInit,
  Input,
  ChangeDetectionStrategy,
} from '@angular/core';

@Component({
  selector: 'app-destination-widget-list',
  templateUrl: './destination-widget-list.component.html',
  styleUrls: ['./destination-widget-list.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class DestinationWidgetListComponent implements OnInit {
  constructor() {}

  @Input() private maxWidgetCount;
  public widgetCountIterator: number[];

  ngOnInit() {
    this.widgetCountIterator = new Array<number>(this.maxWidgetCount).fill(0);
  }
}
