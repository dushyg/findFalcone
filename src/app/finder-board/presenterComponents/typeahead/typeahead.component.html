<div class="typeahead-container" (focusout)="hideResultsList()">
  <div class="typeadhead-wrapper">
    <input
      type="text"
      class="typeAhead"
      placeholder="Type to search..."
      [formControl]="inputTextControl"
      (focus)="showResultsList()"
    />
    <div class="resultList" *ngIf="doShowResults">
      <ul>
        <li class="resultItem" (click)="planetSelectHandler('Select')">
          None
        </li>
        <li
          class="resultItem"
          (click)="planetSelectHandler(planet.name)"
          *ngFor="let planet of filteredSourceArray"
        >
          {{ planet.name }}
        </li>
      </ul>
    </div>
  </div>
</div>
