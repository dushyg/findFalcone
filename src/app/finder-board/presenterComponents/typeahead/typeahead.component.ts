import {
  Component,
  Input,
  Output,
  EventEmitter,
  OnInit,
  OnDestroy,
  OnChanges,
  SimpleChanges,
} from '@angular/core';
import { IPlanet } from 'src/app/finder-board/models/planet';
import { FormControl } from '@angular/forms';
import { Observable } from 'rxjs';
import { takeWhile } from 'rxjs/operators';

@Component({
  selector: 'app-typeahead',
  templateUrl: './typeahead.component.html',
  styleUrls: ['./typeahead.component.scss'],
})
export class TypeaheadComponent implements OnInit, OnChanges, OnDestroy {
  @Input() sourceArray: IPlanet[];
  @Input() resetTypeAhead$: Observable<void>;
  @Output() itemSelected = new EventEmitter<IPlanet>();

  public filteredSourceArray: IPlanet[];
  public inputTextControl = new FormControl('');
  public doShowResults = false;
  private lastSelection = '';
  private isComponentActive = true;

  ngOnInit(): void {
    this.filteredSourceArray = this.sourceArray;

    this.inputTextControl.valueChanges.subscribe(this.performSearch);

    this.resetTypeAhead$
      .pipe(takeWhile(() => this.isComponentActive))
      .subscribe(() => {
        // Need to wrap call to reset in setTimeout because this subscription happens before
        // the sourceArray is updated via inputs. Without setTimeout the reset function will
        // set the filteredSourceArray to old sourceArray.
        setTimeout(() => {
          this.reset();
        }, 0);
      });
  }

  ngOnChanges(changes: SimpleChanges): void {
    console.log('TypeAhead component onchanges', changes);
  }

  private reset() {
    console.log('Typeahead component reset, sourcearray', this.sourceArray);
    this.filteredSourceArray = this.sourceArray;
    this.lastSelection = '';
    this.inputTextControl.setValue('', { emitEvent: false });
  }

  private performSearch = (value) => {
    const trimmedValue = value.trim().toLowerCase();
    if (!trimmedValue) {
      this.filteredSourceArray = this.sourceArray;
    }

    this.filteredSourceArray = this.sourceArray.filter((item) => {
      return item.name.toLowerCase().includes(trimmedValue);
    });
  };

  public planetSelectHandler(planetName) {
    let selectedPlanet = this.filteredSourceArray.find((planet) => {
      return planet.name === planetName;
    });

    if (planetName === 'Select') {
      selectedPlanet = { name: 'Select', distance: 0 } as IPlanet;
      this.lastSelection = '';
      this.inputTextControl.setValue('', { emitEvent: false });
    } else {
      this.lastSelection = selectedPlanet.name;
      this.inputTextControl.setValue(this.lastSelection);
    }

    this.itemSelected.emit(selectedPlanet);
  }

  public showResultsList() {
    this.doShowResults = true;
  }

  public hideResultsList() {
    // focusout event on the wrapper div is fired first,
    // hence we need to call this after some delay so that lastSelection field
    // has latest value
    setTimeout(this.hideResults, 250);
  }

  private hideResults = () => {
    this.doShowResults = false;

    const currentText = this.inputTextControl.value;
    if (currentText === this.lastSelection) {
      return;
    }

    this.inputTextControl.setValue(this.lastSelection, { emitEvent: false });
  };

  ngOnDestroy(): void {
    this.isComponentActive = false;
  }
}
