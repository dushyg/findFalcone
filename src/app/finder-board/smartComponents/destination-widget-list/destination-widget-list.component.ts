import { Component, OnInit } from "@angular/core";
import { FinderFacadeService } from "src/app/finder-board/services/finder-facade.service";

@Component({
  selector: "app-destination-widget-list",
  templateUrl: "./destination-widget-list.component.html",
  styleUrls: ["./destination-widget-list.component.scss"],
})
export class DestinationWidgetListComponent implements OnInit {
  constructor(private finderFacadeService: FinderFacadeService) {}

  public widgetCountIterator: number[];

  ngOnInit() {
    this.widgetCountIterator = new Array<number>(
      this.finderFacadeService.getCountOfWidgetsDisplayed()
    ).fill(0);
  }
}
