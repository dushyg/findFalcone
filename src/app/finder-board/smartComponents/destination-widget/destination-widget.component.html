<div
  class="destinationWidget"
  [attr.widgetId]="widgetId"
  *ngIf="planetListInitialized"
>
  <span>Destination {{ widgetId }}</span>
  <app-typeahead
    [sourceArray]="planetList"
    [resetTypeAhead$]="resetTypeAhead$"
    (itemSelected)="planetSelected($event)"
    [widgetId]="widgetId"
  >
  </app-typeahead>

  <app-vehicle-list
    *ngIf="lastSelectedPlanet != 'Select'"
    [destinationDistance]="destinationDistance"
    [widgetId]="widgetId"
  >
  </app-vehicle-list>
</div>
