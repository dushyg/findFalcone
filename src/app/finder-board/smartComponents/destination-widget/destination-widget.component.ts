import { Component, OnInit, OnDestroy } from "@angular/core";

import { IPlanet } from "src/app/finder-board/models/planet";
import PlanetChange from "src/app/finder-board/models/planetChange";
import { Observable, Subject, combineLatest } from "rxjs";
import { FinderFacadeService } from "src/app/finder-board/services/finder-facade.service";
import { takeWhile, withLatestFrom } from "rxjs/operators";

@Component({
  selector: "app-destination-widget",
  templateUrl: "./destination-widget.component.html",
  styleUrls: ["./destination-widget.component.scss"],
})
export class DestinationWidgetComponent implements OnInit, OnDestroy {
  private static createdWidgetCount: number = 0;
  public widgetId: number;

  public planetList: IPlanet[];

  public destinationDistance: number = 0;

  public lastSelectedPlanet: string = "Select";

  private resetTypeAheadSubject = new Subject<void>();
  public resetTypeAhead$: Observable<
    void
  > = this.resetTypeAheadSubject.asObservable();

  private isComponentActive = true;
  public planetListInitialized = false;
  constructor(private finderFacadeService: FinderFacadeService) {
    // using modulo operator to cycle the widget ids from 1 to max widget count
    this.widgetId =
      (DestinationWidgetComponent.createdWidgetCount %
        this.finderFacadeService.getCountOfWidgetsDisplayed()) +
      1;
    DestinationWidgetComponent.createdWidgetCount++;
  }

  ngOnInit(): void {
    combineLatest(
      this.finderFacadeService.unsearchedPlanets$,
      this.finderFacadeService.isLoading$
    )
      .pipe(takeWhile(() => !this.planetListInitialized))
      .subscribe(([unsearchedPlanets, isLoading]) => {
        if (unsearchedPlanets && !isLoading) {
          this.planetList = unsearchedPlanets;
          this.planetListInitialized = true;
        }
      });

    this.finderFacadeService.planetChangedAction$
      .pipe(
        withLatestFrom(this.finderFacadeService.unsearchedPlanets$),
        takeWhile(() => this.isComponentActive)
      )
      .subscribe(([planetChange, unsearchedPlanets]) => {
        if (planetChange && planetChange.widgetId < this.widgetId) {
          this.planetList = unsearchedPlanets;
          this.clearLastSelection();
        }
      });

    this.finderFacadeService.vehicleChangedAction$
      .pipe(
        withLatestFrom(this.finderFacadeService.unsearchedPlanets$),
        takeWhile(() => this.isComponentActive)
      )
      .subscribe(([vehicleChange, unsearchedPlanets]) => {
        if (vehicleChange && vehicleChange.widgetId < this.widgetId) {
          this.planetList = unsearchedPlanets;
          this.clearLastSelection();
        }
      });
  }

  clearLastSelection(): void {
    this.lastSelectedPlanet = "Select";
    this.resetTypeAheadSubject.next();
  }

  public planetSelected(selectedPlanet: IPlanet) {
    this.destinationDistance = selectedPlanet.distance;

    this.lastSelectedPlanet = selectedPlanet.name;

    this.finderFacadeService.planetChanged(
      new PlanetChange(this.widgetId, selectedPlanet.name)
    );
  }

  ngOnDestroy(): void {
    this.isComponentActive = false;
  }
}
