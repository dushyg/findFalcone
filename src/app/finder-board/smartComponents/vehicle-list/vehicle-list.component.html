<div class="vehicleList">
  <div *ngFor="let vehicle of vehicleList; let i = index">
    <input
      type="radio"
      name="radio_{{ widgetId }}"
      [value]="vehicle.name"
      [ngModel]="lastSelectedVehicle.name"
      [disabled]="
        (vehicle && vehicle.maxDistance < destinationDistance) ||
        vehicle.availNumUnits == 0
      "
      id="vehicle_{{ widgetId }}_{{ i }}"
      (change)="vehicleSelected(vehicle)"
    />&nbsp;
    <label for="vehicle_{{ widgetId }}_{{ i }}">
      {{ vehicle.name + " (" + vehicle.availNumUnits + ")" }}
    </label>

    <br />
  </div>
</div>
