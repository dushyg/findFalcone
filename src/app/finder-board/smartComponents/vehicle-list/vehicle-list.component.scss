.vehicleList {
  display: inline-block;
  text-align: left;
}

input[type="radio"]:disabled {
  cursor: not-allowed;
}

input[type="radio"]:disabled + label {
  cursor: not-allowed;
  text-decoration: line-through;
  color: lightcoral;
}

input[type="radio"]:checked + label {
  font-weight: bold;
  font-style: italic;
  text-decoration: none;
  color: black;
}
