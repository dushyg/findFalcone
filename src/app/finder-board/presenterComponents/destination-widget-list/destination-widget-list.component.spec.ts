import { ComponentFixture, TestBed } from '@angular/core/testing';

import { DestinationWidgetListComponent } from './destination-widget-list.component';
import { DestinationWidgetComponent } from '../../smartComponents/destination-widget/destination-widget.component';
import { createSpyObj } from '../../../testingUtility';
import { TypeaheadComponent } from '../typeahead/typeahead.component';
import { VehicleListComponent } from '../../smartComponents/vehicle-list/vehicle-list.component';
import { FinderFacadeService } from '../../services/finder-facade.service';
import { PlanetsService } from '../../services/planets.service';
import { VehiclesService } from '../../services/vehicles.service';
import { FalconeTokenService } from '../../services/falcone-token.service';
import { Router } from '@angular/router';
import { NO_ERRORS_SCHEMA } from '@angular/core';

describe('DestinationWidgetListComponent', () => {
  let component: DestinationWidgetListComponent;
  let fixture: ComponentFixture<DestinationWidgetListComponent>;
  let planetServiceMock: { [key: string]: jest.Mock<any, any> };
  let vehiclesServiceMock: { [key: string]: jest.Mock<any, any> };
  let apiTokenServiceMock: { [key: string]: jest.Mock<any, any> };
  let routerServiceMock: { [key: string]: jest.Mock<any, any> };

  const apiTokenToBeReturned = { token: 'plmVHX' };

  beforeEach(() => {
    planetServiceMock = createSpyObj(['getAllPlanets']);
    vehiclesServiceMock = createSpyObj(['getAllVehicles']);
    apiTokenServiceMock = createSpyObj(['getFalconeFinderApiToken']);
    routerServiceMock = createSpyObj(['navigate']);

    TestBed.configureTestingModule({
      declarations: [
        DestinationWidgetListComponent,
        DestinationWidgetComponent,
        TypeaheadComponent,
        VehicleListComponent,
      ],
      providers: [
        FinderFacadeService,
        { provide: PlanetsService, useValue: planetServiceMock },
        { provide: VehiclesService, useValue: vehiclesServiceMock },
        { provide: FalconeTokenService, useValue: apiTokenServiceMock },
        { provide: Router, useValue: routerServiceMock },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });
    fixture = TestBed.createComponent(DestinationWidgetListComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
    (expect(fixture) as any).toMatchSnapshot();
  });
});
