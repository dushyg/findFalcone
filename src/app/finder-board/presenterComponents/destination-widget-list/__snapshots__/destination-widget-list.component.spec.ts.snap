// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DestinationWidgetListComponent should create 1`] = `
<app-destination-widget-list
  widgetCountIterator={[Function Array]}
>
  <div
    class="widgetListContainer"
  >
    
    <div
      class="widgetRow"
    >
      <app-destination-widget>
        
      </app-destination-widget>
    </div>
  </div>
</app-destination-widget-list>
`;
